File: transform/Scala/src/main/scala/com/coredf/transform/TransformClient.scala
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.coredf.transform

import org.w3c.dom.Element
import java.io.{ByteArrayInputStream, StringWriter}
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

object TransformClient:
  def JsonParse(text: String): Result =
    try Result.ok(Map("data" -> JsonUtil.parse(text)))
    catch case e: Exception => Result.error(400, e.getMessage)

  def JsonStringify(data: Any, indent: Integer = null): Result =
    try Result.ok(Map("text" -> JsonUtil.stringify(data)))
    catch case e: Exception => Result.error(400, e.getMessage)

  def CsvToRows(text: String, delimiter: String = ","): Result =
    try
      val lines = text.split("\r?\n")
      if lines.isEmpty then Result.ok(Map("rows" -> List.empty[Map[String, String]]))
      else
        val headers = lines(0).split(delimiter, -1)
        val rows = lines.drop(1).filter(_.nonEmpty).map { line =>
          val cols = line.split(delimiter, -1)
          headers.zipWithIndex.map { case (h, i) => h -> (if i < cols.length then cols(i) else "") }.toMap
        }
        Result.ok(Map("rows" -> rows))
    catch case e: Exception => Result.error(400, e.getMessage)

  def RowsToCsv(rows: java.util.List[java.util.Map[String, String]], delimiter: String = ","): Result =
    if rows == null || rows.isEmpty then Result.error(400, "rows must not be empty")
    else try
      val list = rows.toArray.map(_.asInstanceOf[java.util.Map[String, String]])
      val headers = list(0).keySet().toArray.map(_.toString).toList
      val sb = new StringBuilder(headers.mkString(delimiter)).append('\n')
      list.foreach { row =>
        sb.append(headers.map(h => Option(row.get(h)).getOrElse("")).mkString(delimiter)).append('\n')
      }
      Result.ok(Map("text" -> sb.toString))
    catch case e: Exception => Result.error(400, e.getMessage)

  def XmlToDict(text: String): Result =
    try
      val doc = DocumentBuilderFactory.newInstance.newDocumentBuilder.parse(new ByteArrayInputStream(text.getBytes))
      val root = doc.getDocumentElement
      Result.ok(Map("data" -> Map(root.getTagName -> elem(root))))
    catch case e: Exception => Result.error(400, e.getMessage)

  def DictToXml(data: Map[String, Any], rootTag: String = "root"): Result =
    try
      val doc = DocumentBuilderFactory.newInstance.newDocumentBuilder.newDocument()
      val root = doc.createElement(rootTag); doc.appendChild(root)
      data.foreach { case (k, v) => build(root, v, k) }
      val sw = new StringWriter()
      TransformerFactory.newInstance.newTransformer().transform(new DOMSource(doc), new StreamResult(sw))
      Result.ok(Map("text" -> sw.toString))
    catch case e: Exception => Result.error(400, e.getMessage)

  private def elem(node: Element): Any =
    val children = (0 until node.getChildNodes.getLength)
      .map(node.getChildNodes.item)
      .collect { case e: Element => e }
    if children.isEmpty then Option(node.getTextContent).map(_.trim).getOrElse("")
    else
      val out = scala.collection.mutable.LinkedHashMap[String, Any]()
      children.foreach { child =>
        val v = elem(child); val tag = child.getTagName
        if out.contains(tag) then
          val prev = out(tag)
          out(tag) = prev match
            case l: scala.collection.mutable.ArrayBuffer[_] => l += v; l
            case other => scala.collection.mutable.ArrayBuffer(other, v)
        else out(tag) = v
      }
      out.toMap

  private def build(parent: org.w3c.dom.Element, obj: Any, tag: String): Unit =
    val doc = parent.getOwnerDocument
    obj match
      case m: Map[_, _] =>
        val node = doc.createElement(tag); parent.appendChild(node)
        m.foreach { case (k, v) => build(node, v, k.toString) }
      case it: Iterable[_] =>
        it.foreach(v => build(parent, v, tag))
      case other =>
        val node = doc.createElement(tag)
        node.setTextContent(if other == null then "" else other.toString)
        parent.appendChild(node)

File: transform/Java/src/main/java/com/coredf/transform/TransformClient.java
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.coredf.transform;
import org.w3c.dom.*; import javax.xml.parsers.*; import javax.xml.transform.*; import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult; import java.io.*; import java.util.*;

public final class TransformClient {
    private TransformClient() {}
    public static Result JsonParse(String text) {
        try { return Result.ok(Map.of("data", JsonUtil.parse(text))); } catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result JsonStringify(Object data, Integer indent) {
        try {
            String s = JsonUtil.stringify(data);
            return Result.ok(Map.of("text", s));
        } catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result JsonStringify(Object data) { return JsonStringify(data, null); }
    public static Result CsvToRows(String text, String delimiter) {
        try {
            List<Map<String, String>> rows = new ArrayList<>(); String[] lines = text.split("\r?\n");
            if (lines.length == 0) return Result.ok(Map.of("rows", rows));
            String[] headers = lines[0].split(delimiter, -1);
            for (int i = 1; i < lines.length; i++) {
                if (lines[i].isEmpty()) continue; String[] cols = lines[i].split(delimiter, -1);
                Map<String, String> row = new LinkedHashMap<>(); for (int j = 0; j < headers.length; j++) row.put(headers[j], j < cols.length ? cols[j] : "");
                rows.add(row);
            }
            return Result.ok(Map.of("rows", rows));
        } catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result CsvToRows(String text) { return CsvToRows(text, ","); }
    public static Result RowsToCsv(List<Map<String, String>> rows, String delimiter) {
        if (rows == null || rows.isEmpty()) return Result.error(400, "rows must not be empty");
        try {
            List<String> headers = new ArrayList<>(rows.get(0).keySet()); StringBuilder sb = new StringBuilder(String.join(delimiter, headers)).append("\n");
            for (Map<String, String> row : rows) { List<String> vals = new ArrayList<>(); for (String h : headers) vals.add(row.getOrDefault(h, "")); sb.append(String.join(delimiter, vals)).append("\n"); }
            return Result.ok(Map.of("text", sb.toString()));
        } catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result RowsToCsv(List<Map<String, String>> rows) { return RowsToCsv(rows, ","); }
    public static Result XmlToDict(String text) {
        try { Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(new ByteArrayInputStream(text.getBytes()));
            Element root = doc.getDocumentElement(); return Result.ok(Map.of("data", Map.of(root.getTagName(), elem(root)))); }
        catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result DictToXml(Map<String, Object> data, String rootTag) {
        try { Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element root = doc.createElement(rootTag); doc.appendChild(root);
            for (Map.Entry<String, Object> e : data.entrySet()) build(root, e.getValue(), e.getKey());
            Transformer t = TransformerFactory.newInstance().newTransformer();
            StringWriter sw = new StringWriter(); t.transform(new DOMSource(doc), new StreamResult(sw));
            return Result.ok(Map.of("text", sw.toString()));
        } catch (Exception e) { return Result.error(400, e.getMessage()); }
    }
    public static Result DictToXml(Map<String, Object> data) { return DictToXml(data, "root"); }
    private static Object elem(Element node) {
        NodeList children = node.getChildNodes(); boolean hasElem = false;
        for (int i = 0; i < children.getLength(); i++) if (children.item(i) instanceof Element) { hasElem = true; break; }
        if (!hasElem) return node.getTextContent() == null ? "" : node.getTextContent().trim();
        Map<String, Object> out = new LinkedHashMap<>();
        for (int i = 0; i < children.getLength(); i++) {
            if (!(children.item(i) instanceof Element child)) continue;
            Object val = elem(child); String tag = child.getTagName();
            if (out.containsKey(tag)) { Object prev = out.get(tag); if (!(prev instanceof List)) { List<Object> l = new ArrayList<>(); l.add(prev); out.put(tag, l); }
                ((List<Object>) out.get(tag)).add(val); } else out.put(tag, val);
        }
        return out;
    }
    @SuppressWarnings("unchecked")
    private static void build(Element parent, Object obj, String tag) {
        if (obj instanceof Map) { Element node = parent.getOwnerDocument().createElement(tag); parent.appendChild(node);
            for (Map.Entry<String, Object> e : ((Map<String, Object>) obj).entrySet()) build(node, e.getValue(), e.getKey()); }
        else if (obj instanceof List) { for (Object item : (List<?>) obj) build(parent, item, tag); }
        else { Element node = parent.getOwnerDocument().createElement(tag); node.setTextContent(obj == null ? "" : String.valueOf(obj)); parent.appendChild(node); }
    }
}
